src/main/java/org/apache/datasketches/kll/BaseKllSketch.java

```diff
@@ -0,0 +1,324 @@
+   *
+   * Serialized double sketch layout, Empty and Single Item:
+   *  Adr:
+   *      ||    7    |   6   |    5   |    4   |    3   |    2    |    1   |      0       |
+   *  0   || unused  |   M   |--------K--------|  Flags |  FamID  | SerVer | PreambleInts |
+   *      ||                                                               |      8       |
+   *  1   ||----------------------------------data----------------------------------------|
+   */
+
+  /**
+   * The default value of K.
+   */
+  public static final int DEFAULT_K = 200;
+  static final int DEFAULT_M = 8;
+  static final int MIN_K = DEFAULT_M;
+  static final int MAX_K = (1 << 16) - 1; // serialized as an unsigned short
+
+  // Preamble byte addresses
+  static final int PREAMBLE_INTS_BYTE     = 0;
+  static final int SER_VER_BYTE           = 1;
+  static final int FAMILY_BYTE            = 2;
+  static final int FLAGS_BYTE             = 3;
+  static final int K_SHORT                = 4;  // to 5
+  static final int M_BYTE                 = 6;
+  //                                        7 is reserved for future use
+  // SINGLE ITEM ONLY
+  static final int DATA_START_SINGLE_ITEM = 8;
+
+  // MULTI-ITEM
+  static final int N_LONG                 = 8;  // to 15
+  static final int MIN_K_SHORT            = 16; // to 17
+  static final int NUM_LEVELS_BYTE        = 18;
+
+  // FLOAT SKETCH                           19 is reserved for future use in float sketch
+  static final int DATA_START_FLOAT       = 20; // float sketch, not single item
+
+  // DOUBLE SKETCH                          19 to 23 is reserved for future use in double sketch
+  static final int DATA_START_DOUBLE      = 24; // double sketch, not single item
+
+  // Other static values
+  static final byte SERIAL_VERSION            = 1;
+  static final byte SERIAL_VERSION_SINGLE     = 2; // only used to specify the single-item format
+  static final int PREAMBLE_INTS_EMPTY_SINGLE = 2; // for empty and single item
+  static final int PREAMBLE_INTS_FLOAT        = 5; // not empty or single item
+  static final int PREAMBLE_INTS_DOUBLE       = 6; // not empty or single item
+
+  enum Flags { IS_EMPTY, IS_LEVEL_ZERO_SORTED, IS_SINGLE_ITEM }
+
+  /*
+   * Data is stored in items_.
+   * The data for level i lies in positions levels_[i] through levels_[i + 1] - 1 inclusive.
+   * Hence levels_ must contain (numLevels_ + 1) indices.
+   * The valid portion of items_ is completely packed, except for level 0.
+   * Level 0 is filled from the top down.
+   *
+   * Invariants:
+   * 1) After a compaction, or an update, or a merge, all levels are sorted except for level zero.
+   * 2) After a compaction, (sum of capacities) - (sum of items) >= 1,
+   *  so there is room for least 1 more item in level zero.
+   * 3) There are no gaps except at the bottom, so if levels_[0] = 0,
+   *  the sketch is exactly filled to capacity and must be compacted.
+   */
```

Comment:
- An item is an abstraction of the comparable "object" that the sketch is designed to be updated with and compute the quantiles and ranks of after being updated with many such items.  For this base class a unit could be either a float or a double.
- To help the user perform _a priori_ space planning (i.e., how big will the sketch be?) , the user must start with an estimate of N, the largest anticipated number of items that the sketch will see, and the value of K to be used, which determines the error distribution of the sketch.   The user then calls the method _getMaxSerializedSizeBytes(k, N)_, which returns the maximum serialized size of the sketch.  
- Internally, in order to compute this size, the sketch first compute the upper bound on the maximum number of levels required, which is _floor(log2(N))_.  The sketch takes this information and then computes the total number of items that can be retained in a sketch given K, M, numLevels.  Where M is the internal _fixed_ value of the size of the smallest level. 
- To compute this, the sketch then iterates over all levels computing the maximum capacity of each level given K, M, numLevels.  This is a tricky computation that involves numerical folding so that the sketch could handle as many as 60 levels, which would be _extremely_ unlikely.
- The "Total Capacity" then, is the upper bound on the number of items that can be retained in the sketch given  K, M, and the maximum number of levels, which, in turn, is determined from the anticipated N. 
-  In the description, the "sum of capacities" is the "Total Capacity" given K, M, and numLevels (or N).

Reply:
Thanks @leerho - the extra detail that you provided is helpful and answers my question.